### PR TITLE
V2: Add jupter notebook for predicting and training multicomponent

### PR DESCRIPTION
**Pull request by hwpang**
Adding jupter notebooks for predicting single component and training multicomponent

### CODE REVIEW

**Comment by davidegraff** [comment]
Note: should be merged after #501

**Comment by KnathanM** [comment]
My recommendation is to keep the notebook as simple as you can.
Some examples:
1. You explicitly set `num_workers=0` in the dataloader. This is the default and could be removed. 
2. You assign the variable `smiles_column` and only use it once. You could just replace it in `smis = df_test[smiles_column].tolist()` with its value.

I think this is coming along well. Thank you for working on it.

**Comment by hwpang** [comment]
@KnathanM Thanks for the comments! I will work on this today. Was really occupied previously.

**Comment by hwpang** [comment]
During working on this notebook, there are several things I noticed that the `split_data` is currently awkward and can be improved for multi-components.

With https://github.com/chemprop/chemprop/pull/501, a `MulticomponentDataset` dataset contains `K` datasets where `K` is the number of components. These `K` datasets need to be split consistently. The current `split_data` function is not able to do that. I discussed this with @KnathanM and we thought of some ways to do this, and I implemented one of them. Here are the possible options we came up with:

- Option 1: Generalize `split_data` to be able to take in `K` lists of datapoints, use one of them to do the splitting, apply the same splitting to the rest of lists of datapoints. Supply `split_key_molecule` to indicate which one is used for splitting.

- Option 2: Make `split_data` return the split indices. This is the approach I took.

Which way makes more sense?

**Comment by hwpang** [comment]
With these commits I'm able to get MulticomponentMPNN to train. Will work on cleaning up/ prettify the notebook.

**Comment by davidegraff** [comment]
> During working on this notebook, there are several things I noticed that the `split_data` is currently awkward and can be improved for multi-components.
> 
> With #501, a `MulticomponentDataset` dataset contains `K` datasets where `K` is the number of components. These `K` datasets need to be split consistently. The current `split_data` function is not able to do that. I discussed this with @KnathanM and we thought of some ways to do this, and I implemented one of them. Here are the possible options we came up with:
> 
> * Option 1: Generalize `split_data` to be able to take in `K` lists of datapoints, use one of them to do the splitting, apply the same splitting to the rest of lists of datapoints. Supply `split_key_molecule` to indicate which one is used for splitting.
> * Option 2: Deal with this analogous to data scaler. Make `split_data` also return the split indices. If `train_val_test_indices` are provided, split the datapoints. This is the approach I took.
> 
> Which way makes more sense?

This is a good question and something that should be addressed prior to an example PR. As it currently stands, the data/model infrastructure for multicomponent predictions is mostly there in chemprop, but there are still a few wrinkles to iron out. You've identified a couple of these in your PR, so thank you for the great work. I would suggest we close this current PR and open a stack of new PRs aimed at fixing this:
1. fix the simple bugs in the multicomponent code (i.e., your fixes to imports and the `multi.py` module)
2. address the inability for our data splitting to work with our multicomponent data model
3. an example notebook of multicomponent training. You can copy this notebook over from this PR to the final PR so you don't lose the work.

File: chemprop/v2/models/multi.py [hunk 1]
```diff
@@ -16,24 +16,22 @@ def __init__(
...
+        metrics: Iterable[Metric] | None = None,
+        w_t: Tensor | None = None,
         warmup_epochs: int = 2,
         init_lr: float = 1e-4,
         max_lr: float = 1e-3,
         final_lr: float = 1e-4,
     ):
         super().__init__(
-            message_passing,
+            message_passing, # MPNN doesn't take MulticomponentMessagePassing?
```
Comment (davidegraff) [inline chemprop/v2/models/multi.py:29]: the two have largely the same interface, so we can rely on the superclass methods to take care of a bunch of boilerplate even if the objects are not the same

File: chemprop/v2/models/modules/message_passing/multi.py [hunk 2]
```diff
@@ -1,20 +1,21 @@
...
+from lightning import pytorch as pl
 from torch import Tensor, nn
 
 from chemprop.v2.featurizers.molgraph import BatchMolGraph
-from chemprop.v2.models.modules.message_passing.molecule import MessagePassingBlockBase
+from chemprop.v2.models.modules.message_passing.molecule import MessagePassingBlock
 
 
-class MulticomponentMessagePassing(nn.Module):
+class MulticomponentMessagePassing(pl.LightningModule):
```
Comment (davidegraff) [inline chemprop/v2/models/modules/message_passing/multi.py:11]: how is this a `LightningModule`?

**Comment by hwpang** [comment]
> > During working on this notebook, there are several things I noticed that the `split_data` is currently awkward and can be improved for multi-components.
> > With #501, a `MulticomponentDataset` dataset contains `K` datasets where `K` is the number of components. These `K` datasets need to be split consistently. The current `split_data` function is not able to do that. I discussed this with @KnathanM and we thought of some ways to do this, and I implemented one of them. Here are the possible options we came up with:
> > 
> > * Option 1: Generalize `split_data` to be able to take in `K` lists of datapoints, use one of them to do the splitting, apply the same splitting to the rest of lists of datapoints. Supply `split_key_molecule` to indicate which one is used for splitting.
> > * Option 2: Deal with this analogous to data scaler. Make `split_data` also return the split indices. If `train_val_test_indices` are provided, split the datapoints. This is the approach I took.
> > 
> > Which way makes more sense?
> 
> This is a good question and something that should be addressed prior to an example PR. As it currently stands, the data/model infrastructure for multicomponent predictions is mostly there in chemprop, but there are still a few wrinkles to iron out. You've identified a couple of these in your PR, so thank you for the great work. I would suggest we close this current PR and open a stack of new PRs aimed at fixing this:
> 
> 1. fix the simple bugs in the multicomponent code (i.e., your fixes to imports and the `multi.py` module)
> 2. address the inability for our data splitting to work with our multicomponent data model
> 3. an example notebook of multicomponent training. You can copy this notebook over from this PR to the final PR so you don't lose the work.

Thanks for the suggestion! Yeah, I agree that this is the best way forward. I will split this into 3 PRs.

**Comment by hwpang** [comment]
Closing this as it's broken into four PRs: https://github.com/chemprop/chemprop/pull/520, https://github.com/chemprop/chemprop/pull/521, https://github.com/chemprop/chemprop/pull/522, and https://github.com/chemprop/chemprop/pull/523.

**Pull request closed without merging**